FOGL-3026: add support for native Python filter plugins

pintomax opened the pull request:
FOGL-3026: enable discovery of native Python filter plugins

pintomax committed:
FOGL-3026: enable discovery of native Python filter plugins

FOGL-3026: enable discovery of native Python filter plugins

pintomax committed:
Set m_hndl after dlclose

Set m_hndl after dlclose

pintomax committed:
Merge branch 'develop' into FOGL-3026

pintomax committed:
FOGL-3026: track Python initialisation for filters and notification plugins in PluginInterfaceCleanup

FOGL-3026: handle Python initialisation for filters and notification plugins in PluginInterfaceCleanup

praveen-garg commented on python/foglamp/common/plugin_discovery.py:
```diff
@@ -41,11 +42,14 @@ def get_plugins_installed(cls, plugin_type=None, is_config=False):
         elif plugin_type in ['filter', 'notificationDelivery', 'notificationRule']:
             plugins_list = cls.fetch_c_plugins_installed(plugin_type, is_config)
+            plugins_list_filter = cls.fetch_plugins_installed("filter", is_config)
+            plugins_list.extend(plugins_list_filter)
```
@pintomax please check 
https://github.com/foglamp/FogLAMP/pull/1732/files#diff-84ad01acf8d68fab247cedc5084a7a16L49-R51

if plugin_type in request is `filter` which now are expected to discover python based  plugins too `else` part will take care of it. 

**Improved the the unit tests too for new scenario.**

```
tests.unit.python.foglamp.common.test_plugin_discovery.TestPluginDiscovery.test_get_c_notify_plugins_installed

tests.unit.python.foglamp.common.test_plugin_discovery.TestPluginDiscovery.test_get_c_rules_plugins_installed
```
 
these 2 were genuine failure, which helped to identify issue in the code itself!

praveen-garg commented on python/foglamp/common/plugin_discovery.py:
```diff
@@ -41,11 +42,14 @@ def get_plugins_installed(cls, plugin_type=None, is_config=False):
             plugins_list.extend(plugins_list_c_filter)
             plugins_list.extend(plugins_list_c_notify)
             plugins_list.extend(plugins_list_c_rule)
         elif plugin_type in ['filter', 'notificationDelivery', 'notificationRule']:
```
this elif block is only for those types which have only C plugins.

praveen-garg commented on python/foglamp/services/core/api/filters.py:
```diff
@@ -74,8 +76,16 @@
+        #loaded_plugin_info = apiutils.get_plugin_info(plugin_name, dir='filter')
+        try:
+            # Try fetching Python filter
+            plugin_module_path = "{}/python/foglamp/plugins/filter/{}".format(_FOGLAMP_ROOT, plugin_name)
```
@pintomax  https://github.com/foglamp/FogLAMP/commit/46a78659c7dbc91089bfc4dde236085d62bdcc32

Otherwise `foglamp/plugins/installed` API endpoint will result in errors with default install (and until python filter plugin is installed)! 

Make sure to run `tests/system/python/api/test_plugin_discovery.py` and actually
 * api and 
 * e2e system tests

praveen-garg commented on python/foglamp/services/core/api/filters.py:
```diff
@@ -74,8 +76,16 @@
-        # Load C filter plugin info
-        loaded_plugin_info = apiutils.get_plugin_info(plugin_name, dir='filter')
+        # Load C/Python filter plugin info
+        #loaded_plugin_info = apiutils.get_plugin_info(plugin_name, dir='filter')
```
remove these 2 commented statements!

praveen-garg committed:
FOGL-3026 (python) filter dicovery code fixes with unit test  (#1732)

FOGL-3026 code fixes with unit test

pintomax committed:
Added Python inititialisation flag for South service

Added Python inititialisation flag for South service

pintomax commented:
Next commits will add support for multiple instances of native Python plugins (filters and notification rules)

MarkRiddoch commented on C/services/common/plugin_manager.cpp:
```diff
@@ -399,6 +400,10 @@ char		buf[MAXPATHLEN];
 	{
 		pluginHandle = new NotificationPythonPluginHandle(name.c_str(), buf);
 	}
+	else if (type.compare("filter") == 0)
```
plugin_api.h has defines for the filter type strings, it would be nicer to use these.

MarkRiddoch reviewed (dismissed):
Just a minor point, otherwise ok

pintomax committed:
Added defines for plugin_type checks and code format

Added defines for plugin_type checks and code format

pintomax committed:
FOGL-3026: cleanup the right plugin when multiple instances are set

FOGL-3026: cleanup the right plugin when multiple instances are set

pintomax committed:
FOGL-3026: added support for mutiple instances of Python native plugins

FOGL-3026: added support for mutiple instances of Python native plugins

pintomax committed:
Merge branch 'develop' into FOGL-3026

pintomax committed:
Fix for compilation error

Fix for compilation error

pintomax committed:
Update filter_shim for multiple instances

Update filter_shim for multiple instances

pintomax committed:
Merge branch 'develop' into FOGL-3026

pintomax committed:
Fix for filter_shim, multiple instances

Fix for filter_shim, multiple instances

pintomax committed:
FOGL-3026: addition of new Python interpter and its removal

FOGL-3026: addition of new Python interpter and its removal.

For native python plugins with same name we need to start a new Python interpreter, track it's object and remove it.
The Python module object also needs to be removed.

pintomax committed:
Fixes added

Fixes added

pintomax committed:
Fix compilation

Fix compilation

pintomax committed:
FOGL-3026: fixed handling of same plugin name reload

FOGL-3026: fixed handling of same plugin name reload

pintomax committed:
Removed debug message

Removed debug message

pintomax committed:
Merge branch 'develop' into FOGL-3026

pintomax commented:
@AmandeepArora It would be good to test the foglamp-rule-ML-engine-failure with this branch

pintomax commented:
Just created a new repository:

https://github.com/foglamp/foglamp-filter-ema/tree/FOGL-3362

It's the EMA Python native filter plugin

pintomax committed:
Added support for notification delivery Python plugins

Added support for notification delivery Python plugins

MarkRiddoch commented on C/services/common-plugin-interfaces/python/include/python_plugin_common_interface.h:
```diff
@@ -10,18 +10,61 @@
 // This is the map of Python object initialised in each 
 // South, Notification, Filter  plugin interfaces
-static map<string, PyObject*> *pythonModules = new map<string, PyObject *>();;
+static map<string, PythonModule*> *pythonModules = new map<string, PythonModule*>();
```
Is it intended these are shared between all instances of the interface?

MarkRiddoch commented on C/services/common-plugin-interfaces/python/include/python_plugin_common_interface.h:
```diff
@@ -10,18 +10,61 @@
 // it has a different memory address when set/read by
 // PluginInterfaceInit in South, Filter or Notification
+// Only used in plugin_info_fn calls
 static string gPluginName;
```
Is it not possible to wrap this in a class? I'm concerned by the sharing semantics for these static variables.

If there really only should be one in any given process then I think wrapping them in a single class would make it much clearer. Maybe something like a PythonPluginManager class.

MarkRiddoch commented on C/services/common-plugin-interfaces/python/include/python_plugin_common_interface.h:
```diff
@@ -254,71 +363,201 @@ static PLUGIN_INFORMATION *plugin_info_fn()
-		// Failure
-		if (PyErr_Occurred())
+		// Start a new interpreter
+		newInterp = Py_NewInterpreter();
```
Can this fail?

MarkRiddoch commented on C/services/common-plugin-interfaces/python/include/python_plugin_common_interface.h:
```diff
@@ -254,71 +363,201 @@ static PLUGIN_INFORMATION *plugin_info_fn()
+		if (newObj)
+		{
+			// Create a new PythonModule
+			PythonModule* newModule = new PythonModule(newObj,
```
This could fail

MarkRiddoch commented on C/services/filter-plugin-interfaces/python/python_plugin_interface.cpp:
```diff
@@ -161,32 +170,44 @@ static void filter_plugin_reconfigure_fn(PLUGIN_HANDLE handle,
+	if (!pythonHandles)
+	{
+		// Plugin name can not be logged here
+		Logger::getLogger()->error("pythonHandles map is NULL "
```
Can we say something more meaningful to an end user here? Or is this a truly unexpected event?

pintomax replied:
Unexpected event

MarkRiddoch commented on C/services/filter-plugin-interfaces/python/python_plugin_interface.cpp:
```diff
@@ -161,32 +170,44 @@ static void filter_plugin_reconfigure_fn(PLUGIN_HANDLE handle,
 		Logger::getLogger()->fatal("plugin_handle: plugin_ingest(): "
-					   "pModule is NULL for plugin '%s'",
-					   gPluginName.c_str());
+					   "pModule is NULL");
```
We seem to have reduce the information here, is the deliberate?

pintomax replied:
Plugin name cannot be logged if pythonHandles map is not available or handle does not exist un the map

MarkRiddoch commented on C/services/filter-plugin-interfaces/python/python_plugin_interface.cpp:
```diff
@@ -259,62 +281,159 @@ PLUGIN_HANDLE filter_plugin_init_fn(ConfigCategory* config,
+	if (loadModule || reloadModule)
+	{
+		// Start a new interpreter
+		newInterp = Py_NewInterpreter();
```
failure condition?

MarkRiddoch commented on C/services/filter-plugin-interfaces/python/python_plugin_interface.cpp:
```diff
@@ -259,62 +281,159 @@ PLUGIN_HANDLE filter_plugin_init_fn(ConfigCategory* config,
+		// Check for NULL
+		if (newObj)
+		{
+			PythonModule* newModule = new PythonModule(newObj,
```
Failure condition?

MarkRiddoch reviewed (changes requested):
Since this has been a problematic area do you think you could add some documentation, i.e. comments, to explain the policy for creating interpreters, modules etc and when things are shared and when they are not.

Generally a few comments on failure conditions and making error messages more user meaningful if possible.

pintomax committed:
Merge branch 'develop' into FOGL-3026

pintomax committed:
Added checks for Python objects creation

Added checks for Python objects creation

pintomax committed:
Merge branch 'develop' into FOGL-3026

AmandeepArora commented on C/services/common-plugin-interfaces/python/include/python_plugin_common_interface.h:
```diff
@@ -254,71 +363,224 @@ static PLUGIN_INFORMATION *plugin_info_fn()
+		if (it == pythonModules->end())
+		{
+			Logger::getLogger()->debug("plugin_handle: plugin_init(): "
+						   "pModule nof found for plugin '%s': "
```
typo : 'nof'

AmandeepArora commented on C/services/common-plugin-interfaces/python/include/python_plugin_common_interface.h:
```diff
@@ -254,71 +363,224 @@ static PLUGIN_INFORMATION *plugin_info_fn()
+				if (pInfo)
+				{
+					loadPluginType = string(pInfo->type);
+				}
```
else case?

AmandeepArora commented on C/services/common-plugin-interfaces/python/include/python_plugin_common_interface.h:
```diff
@@ -254,71 +363,224 @@ static PLUGIN_INFORMATION *plugin_info_fn()
+				{
+					loadPluginType = string(pInfo->type);
+				}
+			}
```
'else' case is required?

AmandeepArora commented on C/services/common-plugin-interfaces/python/include/python_plugin_common_interface.h:
```diff
@@ -418,19 +701,34 @@ static void plugin_reconfigure_fn(PLUGIN_HANDLE* handle,
+			std::pair<std::map<PLUGIN_HANDLE, PythonModule*>::iterator, bool> ret;
+			ret = pythonHandles->insert(pair<PLUGIN_HANDLE, PythonModule*>
+				((PLUGIN_HANDLE)*handle, currentModule));
+		}
```
'else' case is required here? I think atleast a debug log.

AmandeepArora reviewed (commented):
Looks good. Some minor comments. Also I think many segments of code look quite similar, we should probably create some helper functions and call them rather than repeating these code segments.

pintomax committed:
Review feedback

Review feedback

praveen-garg commented:
Tested this with https://github.com/foglamp/foglamp-filter-ema :

1) Add sinusoid 
a) add ema in **disabled** mode 

![sinusoid_graph](https://user-images.githubusercontent.com/1974003/66915197-0f790380-f036-11e9-9599-54903c9dee18.png)

**expected:** sinusoid reads should not be affected.

b. warnings in syslog

```
Oct 16 12:15:00 raspberrypi s3[9358]: WARNING: FilterPluginInterface currently does not support 'plugin_start', plugin 'ema'
Oct 16 12:15:00 raspberrypi s3[9358]: WARNING: PythonPluginHandle::ResolveSymbol returning NULL for sym=plugin_start
Oct 16 12:15:00 raspberrypi s3[9358]: WARNING: FilterPluginInterface currently does not support 'plugin_start', plugin 'ema'
Oct 16 12:15:00 raspberrypi s3[9358]: WARNING: PythonPluginHandle::ResolveSymbol returning NULL for sym=plugin_start
3. Add ema in disable mode then enable, Getting FATAL in syslog
Oct 16 12:19:44 raspberrypi s3[9358]: ERROR: Called python script method plugin_reconfigure : error while getting result object, plugin 'ema'
Oct 16 12:19:44 raspberrypi s3[9358]: FATAL: logErrorMessage: Error 'NameError("name 'config' is not defined")', plugin 'ema'
```

2. enable ema now, FATAL errors in syslog **and filter does not apply**

```
Oct 16 12:19:44 raspberrypi s3[9358]: ERROR: Called python script method plugin_reconfigure : error while getting result object, plugin 'ema'
Oct 16 12:19:44 raspberrypi s3[9358]: FATAL: logErrorMessage: Error 'NameError("name 'config' is not defined")', plugin 'ema'
```

3. Add ema (fresh) in enabled mode, it works!  **but** 

a) disabling does not change anything. i.e. ema datapoint will contiune to be caluclated for sinusoid
b) same warnings as in 1(b)

**note**..   May be these fixes are needed are on filter side only or core side; or both!


**Reconfig of ema rate and datpoint & regression testing to follow...** /cc @YashTatkondawar

pintomax committed:
Merge branch 'develop' into FOGL-3026

pintomax commented:
@praveen-garg warning messages 1.b) are always  logged. I'll change that

Following your steps on Ubuntu 16 right now

pintomax committed:
Merge branch 'develop' into FOGL-3026

pintomax commented:
@praveen-garg When filter is added, even if not enabled, the readings are converted from C++ to Python and from Python to C++.

The latter conversion sets a new timestamp, so the sinusoid graph is not ok.

I'll check that conversion with @AmandeepArora

MarkRiddoch commented:
> The latter conversion sets a new timestamp, so the sinusoid graph is not ok.

We should never create a new timestamp for a reading once it has been set, regardless of the enabled or disabled state. The timestamp may be converted between languages, but it should always be the same timestamp.

pintomax committed:
FOGL-3026: added check for user_ts in setReadingAttr

FOGL-3026: added check for user_ts in setReadingAttr

pintomax commented:
There was just a missing check for "user_ts" in setReadingAttr()

pintomax commented:
@praveen-garg Could you please check again your above steps?

praveen-garg reviewed (approved):
lgtm! 

https://github.com/foglamp/foglamp-filter-ema/pull/4 for typo and config item order fixes for EMA plugin

pintomax committed:
Merge branch 'develop' into FOGL-3026
